tests/robustness/powerfailure_test.go

```diff
@@ -0,0 +1,197 @@
+//go:build linux
+
+package robustness
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"go.etcd.io/bbolt/tests/dmflakey"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/sys/unix"
+)
+
+// TestRestartFromPowerFailure is to test data after unexpected power failure.
+func TestRestartFromPowerFailure(t *testing.T) {
+	flakey := initFlakeyDevice(t, t.Name(), dmflakey.FSTypeEXT4, "")
+	root := flakey.RootFS()
+
+	dbPath := filepath.Join(root, "boltdb")
+
+	args := []string{"bbolt", "bench",
+		"-work", // keep the database
+		"-path", dbPath,
+		"-count=1000000000",
+		"-batch-size=5", // separate total count into multiple truncation
+	}
+
+	logPath := filepath.Join(t.TempDir(), fmt.Sprintf("%s.log", t.Name()))
+	logFd, err := os.Create(logPath)
+	require.NoError(t, err)
+	defer logFd.Close()
+
+	// FIXME: gofail should support unix socket so that the test cases won't
+	// be conflicted.
+	fpURL := "127.0.0.1:12345"
+
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Stdout = logFd
+	cmd.Stderr = logFd
+	cmd.Env = append(cmd.Env, "GOFAIL_HTTP="+fpURL)
+	t.Logf("start %s", strings.Join(args, " "))
+	require.NoError(t, cmd.Start(), "args: %v", args)
+
+	errCh := make(chan error, 1)
+	go func() {
+		defer close(errCh)
+		errCh <- cmd.Wait()
+	}()
+
+	defer func() {
+		if t.Failed() {
+			logData, err := os.ReadFile(logPath)
+			assert.NoError(t, err)
+			t.Logf("dump log:\n: %s", string(logData))
+		}
+	}()
+
+	time.Sleep(time.Duration(time.Now().UnixNano()%5+1) * time.Second)
+	t.Logf("simulate power failure")
+
+	activeFailpoint(t, fpURL, "beforeSyncMetaPage", "panic")
```

Comment:
I am thinking we should also support forcibly killing the process so that the process can exit at a random point? 

This can be resolved in a followup PR.

Reply:
Yeah. I am thinking about introducing random panic including force-kill. Let me handle this in the follow-up. Thanks.

Reply:
In this test, you inject the failure on device (fs) after the process already terminates. Should we inject the failure (dropWrite) before we terminate(panic) the process?

For the forcibly killing case (we will support it in a followup PR), we do need to inject the failure on device (fs) after the process already terminates.

Reply:
Discuss with @fuweid , let's support more cases in followup PRs
```
Sync times:
t1       t2          t3               x
FS
   f1   f2     f3               f4

if f4 < x
     f4 ~ x

if f4 > x
     t3 ~ x
```

## Use gofailpoint
- Set a huge value for `commit interval`: make sure all data after the last `sync` is lost
- Set proper value for `commit interval`:  make sure part of the data since last `sync` is lost
- Set very small value for `commit interval`: almost no data loss

## forcibly killing the process
- same as above to support different commit interval